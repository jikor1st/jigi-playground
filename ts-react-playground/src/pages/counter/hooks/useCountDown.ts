import { useCallback, useState } from "react";
import useCounter from "./useCounter";
import useInterval from "./useInterval";

export interface UseCountDownOptions {
  countStart: number;
  intervalMS: number;
  countStop?: number;
  isIncrease?: boolean;
}
const useCountDown = ({
  countStart,
  intervalMS,
  countStop,
  isIncrease = false,
}: UseCountDownOptions) => {
  const [isRunning, setIsRunning] = useState(false);

  const { count, increase, decrease, reset } = useCounter(countStart);

  const startCountDown = useCallback(() => setIsRunning(true), []);
  const stopCountDown = useCallback(() => setIsRunning(false), []);

  const resetCountDown = () => {
    stopCountDown();
    reset();
  };

  const countDownCallback = () => {
    if (count === countStop) {
      stopCountDown();
      return;
    }

    if (isIncrease) {
      increase();
    } else {
      decrease();
    }
  };
  useInterval(countDownCallback, isRunning ? intervalMS : null);

  return { count, startCountDown, resetCountDown };
};

export default useCountDown;
