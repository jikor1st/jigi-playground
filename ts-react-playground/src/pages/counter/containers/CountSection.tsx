import useCountDown from "../hooks/useCountDown";
const CountSection = () => {
  const { count, startCountDown, resetCountDown } = useCountDown({
    countStart: 100,
    countStop: 102,
    intervalMS: 1000,
  });
  return (
    <div>
      <strong>{count}</strong>
      <button onClick={startCountDown}>start count down</button>
      <button onClick={resetCountDown}>reset count down</button>
    </div>
  );
};

export default CountSection;
