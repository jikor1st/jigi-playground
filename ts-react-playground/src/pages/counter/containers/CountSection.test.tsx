import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderHook } from "@testing-library/react-hooks";
import useCountDown from "../hooks/useCountDown";

import CountSection from "./CountSection";
describe("<CountSection/>", () => {
  it("스냅샷 매칭", () => {
    const utils = render(<CountSection />);
    expect(utils.container).toMatchSnapshot();
  });
  it("", () => {});
});
