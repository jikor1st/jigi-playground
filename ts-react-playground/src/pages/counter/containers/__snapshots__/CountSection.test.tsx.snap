// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CountSection/> 스냅샷 매칭 1`] = `
<div>
  <div>
    <strong>
      100
    </strong>
    <button>
      start count down
    </button>
    <button>
      reset count down
    </button>
    <p>
      test
    </p>
  </div>
</div>
`;
